### PR TITLE
add WebglCapabilities

### PR DESCRIPTION
**Pull request by gero3**
Added a new class with the capabilities of the webgl renderer.  Discussed in #7072.

### CODE REVIEW

File: src/renderers/webgl/WebGLCapabilities.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    
+    	}
+    
+    	return 'lowp';
+    
+    };
+
+	// GPU capabilities
+	this.precision = parameters.precision !== undefined ? parameters.precision : 'highp',
+	this.logarithmicDepthBuffer = parameters.logarithmicDepthBuffer !== undefined ? parameters.logarithmicDepthBuffer : false;
```
Comment (bhouston) [inline src/renderers/webgl/WebGLCapabilities.js]: This seems to be weird in that it isn't a queried WebGLCapability like everything else.  Why is every other variable queried and this one isn't.

**Comment by gero3** [comment]
@bhouston no problem. If I didn't want feedback on my code, then I wouldn't be working on a open source project :smirk:

File: src/renderers/webgl/WebGLCapabilities.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    
+    			return 'mediump';
+    
+    		}
+    
+    	}
+    
+    	return 'lowp';
+    
+    };
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLCapabilities.js]: This file seems to have a mix of space/tab indents.

Comment (gero3) [inline src/renderers/webgl/WebGLCapabilities.js]: fixed

File: src/renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -1428,11 +1402,11 @@ THREE.WebGLRenderer = function ( parameters ) {
 		var maxLightCount = allocateLights( lights );
 		var maxShadows = allocateShadows( lights );
 		var maxBones = allocateBones( object );
-		var precision = _precision;
+		var precision = capabilities.precision;
 
 		if ( material.precision !== null ) {
 
-			precision = state.getMaxPrecision( material.precision );
+			precision = capabilities.getMaxPrecision( material.precision );
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: I guess we can also remove the `getMaxPrecision` function in `WebGLState`?

Comment (gero3) [inline src/renderers/WebGLRenderer.js]: indeed removed getMaxPrecision from webglstate

File: src/renderers/webgl/WebGLCapabilities.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	this.maxTextureSize = gl.getParameter( gl.MAX_TEXTURE_SIZE );
+	this.maxCubemapSize = gl.getParameter( gl.MAX_CUBE_MAP_TEXTURE_SIZE );
+
+	this.maxAttributes = gl.getParameter( gl.MAX_VERTEX_ATTRIBS );
+	this.maxVertexUniforms = gl.getParameter( gl.MAX_VERTEX_UNIFORM_VECTORS );
+	this.maxVaryings = gl.getParameter( gl.MAX_VARYING_VECTORS );
+	this.maxFragmentUniforms = gl.getParameter( gl.MAX_FRAGMENT_UNIFORM_VECTORS );
+
+
+	this.supportsVertexTextures = this.maxVertexTextures > 0;
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLCapabilities.js]: Should we remove the "supports" from those? As in...

```javascript
this.vertexTextures
this.floatFragmentTexture
this.floatVertexTextures
```

Comment (gero3) [inline src/renderers/webgl/WebGLCapabilities.js]: That makes sense here. but then replacing [this](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L3592-L3645) with a new property called capabilities would make even more sense to me.

Comment (mrdoob) [inline src/renderers/webgl/WebGLCapabilities.js]: Indeed!

File: src/renderers/WebGLRenderer.js [hunk 5]
```diff
@@ -1445,7 +1421,7 @@ THREE.WebGLRenderer = function ( parameters ) {
 		var parameters = {
 
 			precision: precision,
-			supportsVertexTextures: _supportsVertexTextures,
+			supportsVertexTextures: capabilities.supportsVertexTextures,
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: This should now be `capabilities.vertexTextures`.

Comment (gero3) [inline src/renderers/WebGLRenderer.js]: sorry this should be fixed now

Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Are you sure? I don't think the commit went through...

Comment (gero3) [inline src/renderers/WebGLRenderer.js]: My cloud9 workspace didn't want to push the workspace.

File: src/renderers/WebGLRenderer.js [hunk 6]
```diff
@@ -1799,7 +1775,7 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 			}
 
-			if ( _supportsBoneTextures && object.skeleton && object.skeleton.useVertexTexture ) {
+			if ( capabilities.supportsBoneTextures && object.skeleton && object.skeleton.useVertexTexture ) {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: `capabilities.supportsBoneTextures` doesn't exist...

File: src/renderers/WebGLRenderer.js [hunk 7]
```diff
@@ -3508,7 +3484,7 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	function allocateBones ( object ) {
 
-		if ( _supportsBoneTextures && object && object.skeleton && object.skeleton.useVertexTexture ) {
+		if ( capabilities.supportsBoneTextures && object && object.skeleton && object.skeleton.useVertexTexture ) {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: `capabilities.supportsBoneTextures` doesn't exist...

File: src/renderers/webgl/WebGLCapabilities.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+	this.maxAttributes = gl.getParameter( gl.MAX_VERTEX_ATTRIBS );
+	this.maxVertexUniforms = gl.getParameter( gl.MAX_VERTEX_UNIFORM_VECTORS );
+	this.maxVaryings = gl.getParameter( gl.MAX_VARYING_VECTORS );
+	this.maxFragmentUniforms = gl.getParameter( gl.MAX_FRAGMENT_UNIFORM_VECTORS );
+
+
+	this.vertexTextures = this.maxVertexTextures > 0;
+	this.floatFragmentTextures = !! extensions.get( 'OES_texture_float' );
+	this.floatVertexTextures = this.supportsVertexTextures && this.supportsFloatFragmentTextures;
```
Comment (arose) [inline src/renderers/webgl/WebGLCapabilities.js]: no `supports` prefix here

**Comment by mrdoob** [comment]
Many thanks!

**Comment by gero3** [comment]
took a while tough

**Pull request merged**